imports/lib/schemas/User.ts

```diff
@@ -14,56 +13,35 @@ declare module 'meteor/meteor' {
       displayName?: string;
       googleAccount?: string;
       googleAccountId?: string;
-      discordAccount?: t.TypeOf<typeof DiscordAccount>;
+      discordAccount?: DiscordAccountType;
       phoneNumber?: string;
       dingwords?: string[];
     }
   }
 }
 
-export const UserCodec = t.type({
-  username: t.union([t.string, t.undefined]),
-  emails: t.array(t.type({
-    address: t.string,
-    verified: t.boolean,
-  })),
-  createdAt: date,
-  lastLogin: t.union([date, t.undefined]),
-  services: t.union([t.UnknownRecord, t.undefined]),
-  roles: t.union([t.UnknownRecord, t.undefined]),
-  hunts: t.union([t.array(t.string), t.undefined]),
-  displayName: t.union([t.undefined, t.string]),
-  googleAccount: t.union([t.string, t.undefined]),
-  googleAccountId: t.union([t.string, t.undefined]),
-  discordAccount: t.union([DiscordAccount, t.undefined]),
-  phoneNumber: t.union([t.string, t.undefined]),
-  dingwords: t.union([t.array(t.string), t.undefined]),
+// Note: this needs to exactly match the type of Meteor.User, otherwise we will
+// fail typechecking when we use our attachSchema function. Also, because
+// Meteor.users isn't a Model, we can't rely on transforms or defaults in this
+// schema.
+export const User = z.object({
+  _id: stringId,
+  username: z.string().regex(/^[a-z0-9A-Z_]{3,15}$/).optional(),
+  emails: z.object({ address: z.string().regex(Email), verified: z.boolean() }).array().optional(),
+  createdAt: z.date().optional(),
+  lastLogin: z.date().optional(),
+  services: z.any().optional(),
+  profile: z.object({}).optional(),
+  roles: z.record(z.string(), z.string().array()).optional(),
+  hunts: z.string().regex(Id).array().optional(),
+  displayName: z.string().optional(),
+  googleAccount: z.string().optional(),
+  googleAccountId: z.string().optional(),
+  discordAccount: DiscordAccount.optional(),
+  phoneNumber: z.string().optional(),
+  dingwords: z.string().array().optional(),
```

Comment:
For now, I decided to preserve our current behavior, that `MeteorUsers` is just a wrapper around `Meteor.users`. We historically did that to ensure that the schema was applied, although that's less important now that that happens in a standalone file.

However, I think we _may_ want to consider turning `MeteorUsers` into a `Model` object that wraps `Meteor.users` instead. It has the advantage that our application code can only deal with `Model`s, and doesn't need to think about both (e.g. `publishJoinedQuery`, which currently would have to continue accepting both, since we join onto users).

I think that, so long as we guarantee the types lined up, this should be safe (since we'd always be passing `Meteor.User`-typed objects around), and I can't think of any way it would interfere with places that we touch the actual accounts APIs. But still - I think that's an opportunity for future experimentation.